### PR TITLE
feat: bulk delete jobs API & modify Optimus Apply to use bulk delete

### PR DESCRIPTION
**Pull request by ahmadnaufal**
## Summary

1. Add a new bulk delete jobs API as a support for multiple job deletion
2. Modify Optimus Apply to use bulk delete API instead of unary API to handle this case

### CODE REVIEW

File: core/job/service/job_service.go [hunk 1]
```diff
@@ -1999,3 +1993,143 @@ func (j *JobService) getDownstreamJobs(ctx context.Context, projectName tenant.P
...
+
+	err = validateDeleteJob(downstreamsPerLevel, toDeleteSpecMap)
+	if err != nil {
+		return nil, err
+	}
+
+	// flatten downstreams per level into a single list,
+	// with direct downstreams will be put first & the leaf downstreams at the end
+	downstreams := []*job.Downstream{}
+	for _, dss := range downstreamsPerLevel {
```
Comment (deryrahman) [inline core/job/service/job_service.go:2060]: can we rename it with more readable variable? eg. `currentDownstreams`

File: core/job/service/job_service.go [hunk 2]
```diff
@@ -1999,3 +1993,143 @@ func (j *JobService) getDownstreamJobs(ctx context.Context, projectName tenant.P
...
+		toDeleteJobs = append(toDeleteJobs, existingJob)
+	}
+
+	toDeleteMap := toDeleteJobs.GetNameMap()
+	toDeleteSpecMap := toDeleteJobs.GetFullNameToSpecMap()
+
+	alreadyDeleted := map[job.FullName]bool{}
+	deletedJobs := job.Jobs{}
+	for _, toDeleteJob := range toDeleteJobs {
+		deletionTracker, deletedJobNames := j.resolveJobAndDownstreamsDeletion(ctx, toDeleteJob, toDeleteSpecMap, alreadyDeleted)
```
Comment (deryrahman) [inline core/job/service/job_service.go:2018]: can we also skip downstream deletion if `toDeleteJob` is already been proceeded? by checking on `deletionTrackerMap`

File: core/job/service/job_service.go [hunk 3]
```diff
@@ -1999,3 +1993,143 @@ func (j *JobService) getDownstreamJobs(ctx context.Context, projectName tenant.P
...
+
+		toDeleteJobs = append(toDeleteJobs, existingJob)
+	}
+
+	toDeleteMap := toDeleteJobs.GetNameMap()
+	toDeleteSpecMap := toDeleteJobs.GetFullNameToSpecMap()
+
+	alreadyDeleted := map[job.FullName]bool{}
+	deletedJobs := job.Jobs{}
+	for _, toDeleteJob := range toDeleteJobs {
```
Comment (deryrahman) [inline core/job/service/job_service.go:2017]: can we make the job deletion check in upfront before actually delete it? if we have set of job deletion `S` and it's not safe to delete, and we have subset of `S`, eg. `S1` and it's safe to delete, then job deletion can be partially done (`S1`), but not whole(`S`). as a consequences, I believe we can't retry the pipeline without manual intervention

File: client/cmd/apply/apply.go [hunk 4]
```diff
@@ -185,17 +186,46 @@ func (c *applyCommand) printFailed(namespaceName, operation, kind, name, cause s
...
+	}
+
+	allJobs := []string{}
+	for _, jobToDelete := range request.Jobs {
+		allJobs = append(allJobs, jobToDelete.JobName)
+	}
+
+	response, err := client.BulkDeleteJobs(ctx, request)
+	if err != nil {
+		c.printFailedAll("bulk-delete", "job", strings.Join(allJobs, ","), err.Error())
```
[the remark this answers is not in the document]
Comment (ahmadnaufal) [reply client/cmd/apply/apply.go:209]: will update

File: client/cmd/apply/apply.go [hunk 5]
```diff
@@ -185,17 +186,46 @@ func (c *applyCommand) printFailed(namespaceName, operation, kind, name, cause s
...
 			continue
 		}
-		c.printSuccess(request.NamespaceName, "delete", "job", request.GetJobName())
-		deletedJobs = append(deletedJobs, request.GetJobName())
+
+		if result.GetSuccess() {
+			c.printSuccess(jobToDelete.NamespaceName, "bulk-delete", "job", jobToDelete.JobName)
+			deletedJobs = append(deletedJobs, jobToDelete.JobName)
+		} else {
+			c.printFailed(jobToDelete.NamespaceName, "bulk-delete", "job", jobToDelete.JobName, result.GetMessage())
```
[the remark this answers is not in the document]
Comment (ahmadnaufal) [reply client/cmd/apply/apply.go:225]: I'll put up 2 examples for this:
1. If deletion fails due to Optimus DB returns error for some of the jobs: users can retry the deletion attempt through CLI & updated plan after apply
2. If deletion fails due to some of the jobs have dependencies that aren't deleted: users can create a new MR to delete the dependencies, then retry the previous CI which deletes the upstream job

File: core/job/service/job_service.go (hunk 1 above)
Comment (ahmadnaufal) [reply core/job/service/job_service.go:2060]: agree

File: core/job/service/job_service.go (hunk 2 above)
Comment (ahmadnaufal) [reply core/job/service/job_service.go:2018]: see below answer

File: core/job/service/job_service.go (hunk 3 above)
Comment (ahmadnaufal) [reply core/job/service/job_service.go:2017]: After separate discussion, we've decided to separate the steps to:
1. validation, and
2. the actual deletion.
This require code refactor, will timebox this until 2PM. Else, will go forward with the current approach

**Review by deryrahman — APPROVED** [review]
LGTM

**Pull request merged**